pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/AuthorizationService.java

```diff
@@ -183,7 +183,15 @@ public CompletableFuture<Boolean> canProduceAsync(TopicName topicName, String ro
         if (!this.conf.isAuthorizationEnabled()) {
             return CompletableFuture.completedFuture(true);
         }
-        return provider.allowTopicOperationAsync(topicName, role, TopicOperation.PRODUCE, authenticationData);
+
+        return isSuperUserOrAdmin(topicName.getTenant(), role, authenticationData)
```

Comment:
I do not think we should move in this direction. I think our target design should be to let the `AuthorizationProvider` implementation decide what level of permission is required. This line here removes that abstraction from the `AuthorizationProvider` and puts it in the `AuthorizationService`. I concede that there are some places where we use these checks outside of the `AuthorizationProvider`, however, those are legacy details that we have to support until we find a way to move them into the `AuthorizationProvider`.

Reply:
> I think our target design should be to let the `AuthorizationProvider` implementation decide what level of permission is required

We have `isTenantAdmin` and `isSuperUser` methods in the `AuthorizationProvider`. The user just overrides that to decide what level of permission is required. When `isTenantAdmin` and `isSuperUser` return `false`, the `AuthorizationService` call to the `AuthorizationProvider`.

There is another reason that I think plugin authors should not pay attention to `isTenantAdmin` and `isSuperUser` details.

Reply:
> We have `isTenantAdmin` and `isSuperUser` methods in the `AuthorizationProvider`.

That is a fair point. I think I've been operating under the assumption that these are a bit of an anti-pattern with the idea that it is better to give users the ability to configure granular permissions.

> I think plugin authors should not pay attention to `isTenantAdmin` and `isSuperUser` details.

I agree with this, and this is part of why I do not think we should create new dependencies on these methods. When I added `isSuperUserOrAdmin` recently, I did so because it was necessary to consolidate existing logic. I made it private in an attempt to prevent new usage, but of course that wouldn't stop the `AuthorizationService` from using it.

Reply:
> When `isTenantAdmin` and `isSuperUser` return `false`, the `AuthorizationService` call to the `AuthorizationProvider`.

My primary objection to moving this logic out of the `AuthorizationProvider` is that implementations have almost certainly already accounted for the fact that this method (and all of the similar ones modified in this PR) do not themselves check if the client is a super user or a tenant admin. As such, we could be creating unnecessary overhead by moving this logic out of the `AuthorizationProvider` and into the `AuthorizationService`.

These are my initial thoughts. I'll continue to think about this more. Thanks for the discussion!